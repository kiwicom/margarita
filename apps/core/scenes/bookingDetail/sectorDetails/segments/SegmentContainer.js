// @flow

import * as React from 'react';
import { Touchable, Icon, StyleSheet } from '@kiwicom/universal-components';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';
import { Animated, Platform } from 'react-native';
import { graphql, createFragmentContainer } from '@kiwicom/margarita-relay';
import { BookingTypeRenderer } from '@kiwicom/margarita-components';

import SegmentMap from './SegmentMap';
import type { SegmentContainer_data as BookingType } from './__generated__/SegmentContainer_data.graphql';
import SectorsListOneWay from '../SectorsListOneWay';
import SectorsListReturn from '../SectorsListReturn';
import SectorsListMulticity from '../SectorsListMulticity';

type Props = {|
  +data: ?BookingType,
|};

type State = {|
  +expandSegments: boolean,
  +isDisabled: boolean,
|};

export class SegmentContainer extends React.Component<Props, State> {
  constructor() {
    super();

    this.state = {
      expandSegments: false,
      isDisabled: false,
    };

    this.spinValue = new Animated.Value(0);
  }

  spinValue: Animated.Value;

  toggleExpanded = () => {
    this.setState({ isDisabled: true }, () => {
      Animated.timing(this.spinValue, {
        toValue: this.state.expandSegments ? 0 : 1,
        duration: 250,
        useNativeDriver: true,
      }).start(() => {
        this.setState(state => ({
          expandSegments: !state.expandSegments,
          isDisabled: false,
        }));
      });
    });
  };

  render() {
    const spin = this.spinValue.interpolate({
      inputRange: [0, 1],
      outputRange: ['0deg', '180deg'],
    });

    return (
      <>
        <Touchable
          onPress={this.toggleExpanded}
          disabled={this.state.isDisabled}
        >
          <Animated.View
            style={[
              styles.iconWrapper,
              {
                transform: [{ rotate: spin }, { perspective: 1000 }],
              },
            ]}
          >
            <Icon
              name="chevron-down"
              color={defaultTokens.paletteProductNormal}
            />
          </Animated.View>
        </Touchable>
        {this.state.expandSegments && (
          <>
            <BookingTypeRenderer
              type={this.props.data?.__typename}
              oneWayComponent={<SectorsListOneWay data={this.props.data} />}
              returnComponent={<SectorsListReturn data={this.props.data} />}
              multicityComponent={
                <SectorsListMulticity data={this.props.data} />
              }
            />
            {Platform.OS !== 'web' && <SegmentMap data={this.props.data} />}
          </>
        )}
      </>
    );
  }
}

const styles = StyleSheet.create({
  iconWrapper: {
    flex: 1,
    alignSelf: 'center',
  },
});

export default createFragmentContainer(SegmentContainer, {
  data: graphql`
    fragment SegmentContainer_data on BookingInterface {
      __typename
      ...SectorsListOneWay_data
      ...SectorsListReturn_data
      ...SectorsListMulticity_data
      ...SegmentMap_data
    }
  `,
});
