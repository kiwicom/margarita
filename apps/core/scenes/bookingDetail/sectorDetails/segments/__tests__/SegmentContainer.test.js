// @flow

import * as React from 'react';
import { shallow } from 'react-native-testing-library';

import { SegmentContainer } from '../SegmentContainer';

it('renders', () => {
  expect(shallow(<SegmentContainer data={null} />)).toMatchSnapshot();
});
