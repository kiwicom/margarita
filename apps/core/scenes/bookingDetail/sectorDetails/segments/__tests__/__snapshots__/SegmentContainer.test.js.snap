// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
Object {
  "output": <React.Fragment>
    <Touchable
      accessibilityComponentType="button"
      accessibilityTraits="button"
      borderlessRipple={false}
      disabled={false}
      onPress={[Function]}
      rippleColor="rgba(0, 0, 0, 0.32)"
    >
      <AnimatedComponent
        style={
          Array [
            Object {
              "alignSelf": "center",
              "flex": 1,
            },
            Object {
              "transform": Array [
                Object {
                  "rotate": "0deg",
                },
                Object {
                  "perspective": 1000,
                },
              ],
            },
          ]
        }
      >
        <Icon
          color="#00a991"
          name="chevron-down"
        />
      </AnimatedComponent>
    </Touchable>
  </React.Fragment>,
}
`;
